### ISSUE TITLE
I think the lua file limit should be raised

### ISSUE THREAD

**Issue by clarkxl** [Facepunch/garrysmod-issues#4870]
Details: a previous issue (same topic) ended with the lua file limit being raised to 8096, and if you play gmod for a while and get some decent sized addons you can hit that limit within about 3-6 months, and past that point new addons dont work (due to no new addons being able to load new lua files) and this could also caused updated addons (when a addon updates) to not work as the limit was already hit using other addons. On the other issue on this topic it was said that the limit couldnt be removed, so just wondering, is it possible to raise it higher?

How to reproduce:
have a good amount of addons

**Comment by viral32111** [comment]
I think you and me have different definitions of "a *good* amount of addons"

**Comment by shadowscion** [comment]
> 
> 
> > I think you and me have different definitions of "a _good_ amount of addons"
> 
> well a _good_ few add-ons usually has a few big dependencie, which can fill up that space (e.g wiremod) (I AM NOT BLAMING ANYONE FOR ANYTHING /g)

Wiremod has around 400 lua files and that's probably the most you're going to see in a worthwhile addon

**Comment by RC-3197** [comment]
I'm going to bump this by saying that yes, the LUA limit is lower than it needs to be to keep up with the large volume of good mods coming out of the workshop. Let me put it to you this way, I got this game within the past year and I've already hit the LUA limit, if I did that there's no doubt many other people have too. So in a game about the endless possibilities of modding can we increase this restriction to something like 10,000 flat (that would be more than enough for me) or even something crazy like 16,192 so you never have to hear a post like this again? Anyways, this game is great, I just wish I could have more mods.

**Comment by FlorianLeChat** [comment]
> the LUA limit is lower than it needs to be to keep up with the large volume of good mods coming out of the workshop

Having the maximum number of addons on your game isn't a competition you know. Moreover, exceeding a certain limit can severely degrade your game experience and overall performance. I don't understand why you want to keep increasing this limit when most of the technical problems players report (on Garry's Mod Discord) are caused by too many addons breaking/dysfunctioning some basic game functionalities. Obviously, servers which generate custom content are likely to be **less** bothered by this limit than servers which add **several hundred** addons to their Workshop collection for no reason.

**Comment by RC-3197** [comment]
> Having the maximum number of addons on your game isn't a competition you know.

Did I say it was some sort of competition? No, I'm just saying there are so many good mods coming out of the workshop lately that the limit is not enough to allow us to install them.

> Moreover, exceeding a certain limit can severely degrade your game experience and overall performance.

That's a sacrifice I'm willing to make, I'm sure other people are too, and quite honestly the limit is only about 100 away from what I need to get all the mods I have running at once. Taking the choice away from the user in a game all about user choice, doesn't make any sense.

>I don't understand why you want to keep increasing this limit when most of the technical problems players report (on Garry's Mod Discord) are caused by too many addons breaking/dysfunctioning some basic game functionalities.

It's the user's responsibility to ensure the mods are compatible with each other. And the number of mods really doesn't affect this issue at all, even if you had only 2 mods that don't play nice with eachother, you'll have problems with the game's functionalities.
Not to mention there's 2 workaround's ([from this page](https://github.com/Facepunch/garrysmod-issues/issues/1058#issue-32462836)) with the sacrifice of not being able to update your mods:

> 1. merge lua files together (stupid)
> 2. move lua files from addons folder to garrysmod/lua folder (stupid)

So it's not like it's not physically possible to launch additional LUA files, it's just limited by count of the string table.

> Obviously, servers which generate custom content are likely to be less bothered by this limit than servers which add several hundred addons to their Workshop collection for no reason.

I couldn't care less what a server does, that's their choice. I'll play the game the way I want to and they'll play the game they way they want to. The problem is that the limit is taking away people's ability to choose.

Bottom line is, this limiting number is hard coded by the exe, it's a number limiting a string table that has been changed before. It's been 6 years since it was last changed and the workshop has grown in both quality and quantity. It's time for this number to grow with it.

**Comment by FlorianLeChat** [comment]
> Did I say it was some sort of competition? No, I'm just saying there are so many good mods coming out of the workshop lately that the limit is not enough to allow us to install them.

Don't make me believe that you use all the addons you have installed, it's impossible to imagine.

> That's a sacrifice I'm willing to make, I'm sure other people are too, and quite honestly the limit is only about 100 to 200 away from what I need to get all the mods I have running at once. Taking the choice away from the user in a game all about user choice, doesn't make any sense.

As Rubat said on another issue, if Garry originally put a hardcoded limit in the game it's simply because he has a reason either at the engine technical level or to avoid abuse. Personally I find it pretty normal and logical.

> It's the user's responsibility to ensure the mods are compatible with each other. And the number of mods really doesn't affect this issue at all, even if you had only 2 mods that don't play nice with eachother, you'll have problems with the game's functionalities.

It' s definitely the responsibility of the players to make sure their addons work properly and without conflict with others, but it still won't solve the problem of players complaining about the game being slow or broken because of certain addons, and it could only make the situation worse if this limit is raised further. Trusting players at this level is just unthinkable and will not help.  

Also, it's obvious that the number of addons has an impact on the amount of malfunctions, especially on the performance level. What is the point of a game not having this kind of limit and having players install addons until their game can no longer start or is unstable? The developers obviously have to set some limits to avoid just this kind of problem.

> I couldn't care less what a server does, that's their choice. I'll play the game the way I want to and they'll play the game the way they want to. The problem is that the limit is taking away people's ability to choose.

A good majority of players run Garry's Mod to play on public servers, so it's obvious that some of them will reach this limit sooner or later. I'm surprised that I've never heard this kind of feedback despite the fact most DarkRP servers are overloaded with addons. Which makes me think the limit is for the moment quite reasonable and that this problem is affecting a minority.

> Bottom line is, this limiting number is hard coded by the exe, it's a number limiting a string table that has been changed before. It's been 6 years since it was last changed and the workshop has grown in both quality and quantity. It's time for this number to grow with it.

The addons quality has no influence on the amount of Lua files required, it all depends on the architecture of the scripts. Moreover, this paragraph reinforces my idea that you don't need to have or install so many addons. I wonder how many addons you have currently installed?

**Comment by clarkxl** [comment]
Based on what ive read here, its just changing a number, so why not just make it basically unlimited like RC said? (by putting it to the integer limit)

**Comment by robotboy655** [comment]
I have bumped the max files count to 65k in singleplayer on `dev` beta. Let's see if that breaks anything.

**Comment by robotboy655** [comment]
To be clear, limits are not there to fuck you over or to "limit you", it's there to prevent catastrophic failures, such as crashes, excessive bandwidth/CPU/memory usage, etc.

**Comment by RC-3197** [comment]
> I have bumped the max files count to 32k in **singleplayer only** on `dev` beta. Let's see if that breaks anything.

I modified my game through reverse engineering to have a client_lua_files string table with a capacity of 16,384 (matching the soundprecache string table) a couple of days ago (TBH, I didn't think you'd raise it as much as you did but changing it from 0x2000 to 0x8000 makes sense and is much better). Do you want me to test it in multiplayer (obviously with both the client and the host having the modification) to see if it works okay? I have a hunch that it probably will work (since it seems to work okay in single player) as long as I get all the untested mods I installed to exceed the string table limit to cooperate. Also my sibling has an absolute potato of a PC (yet it can still handle just as many mods as I can), so it would be a really good test of what would happen with a low end pc when this number is changed.

**Comment by RC-3197** [comment]
> > > I have bumped the max files count to 32k in **singleplayer only** on `dev` beta. Let's see if that breaks anything.
> > 
> > 
> > Quick question, I modified my game through hex editing to have a client_lua_files string table with a capacity of 16,384 (matching the soundprecache string table) a couple of days ago (I didn't think you'd raise it as much as you did, but changing it from 0x2000 to 0x8000 (32k) is way better). Do you want me to test 32k in peer-to-peer or LAN multiplayer (obviously with both the client and the host having the 32k modification) to see if it works? I have a hunch that it probably will work, since it seems to work okay in single player. Also, my sibling has an absolute potato of a PC (yet it can still handle just as many mods as I can), so it would be a really good test of what would happen with a low end pc when this number is changed.
> > Edit: In my testing so far, I can successfully host a local server with both the 16k (if you wanted to down scale for MP) and 32k LUA table capacity. I haven't implemented the modification on the second computer yet because it's currently being used. Because of that, I can't test the client connection yet. So instead I'm currently trying to figure out what the maximum string table capacity that I can locally host with (just for fun). So far, I haven't gotten anything past 32k to work yet (which is fine, you're more likely to fill your hard drive before you hit 32k LUA files).
> > Edit: Now I know why it can't use a larger number, it's a short (int 16) not a word (int 32). Now I feel like an idiot for accidentally setting the limit to 0 over and over again (because the int 16 field was empty) and expecting it to work. And since it must be a number that's a base of 2, 32k is the maximum. **Which means you've literally set it to the maximum number the game can physically accept (with this data type) which is awesome! Thank you!** Now I just have to wait until the other computer is unoccupied, so I can test joining a server with the 32k table capacity.
> > Edit: **I am happy to inform you that 32k can work in multiplayer (at least LAN and Peer-to-peer) if implemented! Thank you so much!**
> 
> Can you tell me how to do it please?

Hi! The process involves a modification to one of the Garry's Mod files using a hex editor, because its a modification of a game binary, I can't explain it to you without permission from a developer first. But I can tell you a few things about how it operates, so you can decide whether the trouble is worth it or not:
1. Every client must have the modified file because the LUA limit is client side, otherwise they'll hit the LUA limit when trying to connect to a modified host. I haven't tested the other way around, if I were to guess it'd either disconnect you or adopt the host's LUA limit.
2. Don't go on public servers, because I don't know if modifying the file will affect your steam account standing or not. It's best to not find out the hard way. So just stick to LAN (or peer to peer only if you can't get LAN working) with friends who have made the modification.
3. If you have bad internet like me, you'll have to crank up the disconnect timeouts so people can load the massive number of mods. It can be tricky to get everyone to connect sometimes, which is probably the reason they don't want to make this change in multiplayer quite yet. Know your mod list well, because if its a laggy combination of mods (ex lots of unstable, outdated, or incompatible mods) it will be harder to connect, if it's a very stable collection of mods you shouldn't have many issues, and after loading it will run like normal. Then again, this could just be my really bad internet, in my house it's easier to connect through the internet than it is to connect over LAN in nearly every game.
4. You need to modify the file every time Garry's Mod gets an update, and sometimes the location where the hexadecimal value is changes. I've come up with a consistent way to find it, but that involves using more than just a hex editor. The process itself is quite simple, but learning how to do it might not be worth it to you.

It's definitely possible to make the modification, it just takes a bit of work, and being willing to not connect to or host public servers. So the best way to get it in the game would be to encourage the developers make it a toggleable option in the settings, one with a disclaimer that says something like "loading more than 8000 LUA files in multiplayer is an experimental feature that can lead to instability, are you sure you want to enable this option?". That way, users can choose whether they want to troubleshoot their mod collection or not. The real problem with Garry's Mod instability this isn't the number of LUA files, it's really mod to mod compatibility, out of date mods, and unoptimized mods. I've subscribed to 1460 mods in  Garry's Mod, far exceeding the 8000 file LUA limit, and I can start a basic LAN game with my friends without running into any problems that weren't caused by one of the 3 issues I mentioned earlier.

**Comment by RylostKy** [comment]
> > > > I have bumped the max files count to 32k in **singleplayer only** on `dev` beta. Let's see if that breaks anything.
> > > 
> > > 
> > > Quick question, I modified my game through hex editing to have a client_lua_files string table with a capacity of 16,384 (matching the soundprecache string table) a couple of days ago (I didn't think you'd raise it as much as you did, but changing it from 0x2000 to 0x8000 (32k) is way better). Do you want me to test 32k in peer-to-peer or LAN multiplayer (obviously with both the client and the host having the 32k modification) to see if it works? I have a hunch that it probably will work, since it seems to work okay in single player. Also, my sibling has an absolute potato of a PC (yet it can still handle just as many mods as I can), so it would be a really good test of what would happen with a low end pc when this number is changed.
> > > Edit: In my testing so far, I can successfully host a local server with both the 16k (if you wanted to down scale for MP) and 32k LUA table capacity. I haven't implemented the modification on the second computer yet because it's currently being used. Because of that, I can't test the client connection yet. So instead I'm currently trying to figure out what the maximum string table capacity that I can locally host with (just for fun). So far, I haven't gotten anything past 32k to work yet (which is fine, you're more likely to fill your hard drive before you hit 32k LUA files).
> > > Edit: Now I know why it can't use a larger number, it's a short (int 16) not a word (int 32). Now I feel like an idiot for accidentally setting the limit to 0 over and over again (because the int 16 field was empty) and expecting it to work. And since it must be a number that's a base of 2, 32k is the maximum. **Which means you've literally set it to the maximum number the game can physically accept (with this data type) which is awesome! Thank you!** Now I just have to wait until the other computer is unoccupied, so I can test joining a server with the 32k table capacity.
> > > Edit: **I am happy to inform you that 32k can work in multiplayer (at least LAN and Peer-to-peer) if implemented! Thank you so much!**
> > 
> > 
> > Can you tell me how to do it please?
> 
> Hi! The process involves a modification to one of the Garry's Mod files using a hex editor, because its a modification of a game binary, I can't explain it to you without permission from a developer first. But I can tell you a few things about how the modification operates, so you can decide whether the trouble is worth it or not:
> 
> 1. Every client must have the modified file because the LUA limit is client side, otherwise they'll hit the LUA limit when trying to connect to a modified host. I haven't tested the other way around, if I were to guess it'd either disconnect you or adopt the host's LUA limit.
> 2. Don't go on public servers, because I don't know if modifying the file will affect your steam account standing or not. It's best to not find out the hard way. So just stick to LAN (or peer to peer only if you can't get LAN working) with friends who have made the modification.
> 3. If you have bad internet like me, you'll have to crank up the disconnect timeouts so people can load the massive number of mods. It can be tricky to get everyone to connect sometimes, which is probably the reason they don't want to make this change in multiplayer quite yet. Know your mod list well, because if its a laggy combination of mods (ex lots of unstable, outdated, or incompatible mods) it will be harder to connect, if it's a very stable collection of mods you shouldn't have many issues, and after loading it will run like normal. Then again, this could just be my really bad internet, in my house it's easier to connect through the internet than it is to connect over LAN in nearly every game.
> 4. You need to modify the file every time Garry's Mod gets an update, and sometimes the location where the hexadecimal value is changes. I've come up with a consistent way to find it, but that involves using more than just a hex editor. The process itself is quite simple, but learning how to do it might not be worth it to you.
> 
> It's definitely possible to make the modification, it just takes a bit of work, and being willing to not connect to or host public servers. So the best way to get it in the game would be to encourage the developers make it a toggleable option in the settings, one with a disclaimer that says something like "loading more than 8000 LUA files in multiplayer is an experimental feature that can lead to instability, are you sure you want to enable this option?". That way, users can choose whether they want to troubleshoot their mod collection or not. The real problem with Garry's Mod instability this isn't the number of LUA files, it's really mod incompatibility, out of date mods, and unoptimized mods that cause lag. I've subscribed to 1460 mods in Garry's Mod, far exceeding the 8000 file LUA limit, and I can start a basic LAN game with my friends without running into any problems that weren't caused by one of the 3 issues I mentioned earlier.

Hey just to know, you can atleast say where i can find the file to do that modification ?

**Comment by funkynuggets** [comment]
Would also be intrested in that.
